[android] test_app example linking to pytorch_android aar content

IvanKobzarev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #40176 [android] gradle version update
* **#39587 [android] test_app example linking to pytorch_android aar content**

Example of using direct linking to pytorch_jni library from aar and updating android/README.md with the tutorial how to do it.

Adding `nativeBuild` dimension to `test_app`, using direct aar dependencies, as headers packaging is not landed yet, excluding `nativeBuild` from building by default for CI.

Additional change to `scripts/build_pytorch_android.sh`:

Skipping clean task here as android gradle plugin 3.3.2 exteralNativeBuild has problems with it when abiFilters are specified.

Will be returned back in the following diffs with upgrading of gradle and android gradle plugin versions.

Differential Revision: [D22118945](https://our.internmc.facebook.com/intern/diff/D22118945)

IvanKobzarev committed:
Update on "[WIP][android] test_app example linking to pytorch_android aar content"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[android] test_app example linking to pytorch_android aar content"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[android] test_app example linking to pytorch_android aar content"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[android] test_app example linking to pytorch_android aar content"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[android] test_app example linking to pytorch_android aar content"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[android] test_app example linking to pytorch_android aar content"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[android] test_app example linking to pytorch_android aar content"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[android] test_app example linking to pytorch_android aar content"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[android] test_app example linking to pytorch_android aar content"




[ghstack-poisoned]

ljk53 commented on android/test_app/app/build.gradle:
```diff
@@ -25,17 +31,37 @@ android {
     }
+    signingConfigs {
+        config {
+            storeFile file("testapp.keystore")
```
curious: what's the common practice to handle keystore for demo/test app? I found this but not sure whether it's worth the trouble for test app: https://stackoverflow.com/questions/44911547/where-to-store-android-keystore-file-for-cirlceci-build

IvanKobzarev replied:
Yes, we can do it more securely for CircleCI. That keystore is completely dummy, there is no 'PyTorch' and 'Facebook' inside it. 
I think if somebody will start signing his app with this keystore there should not be any problems.

ljk53 commented on android/test_app/app/src/main/cpp/pytorch_testapp_jni.cpp:
```diff
@@ -0,0 +1,74 @@
+  ALOGI("%s %s", m, os.str().c_str());
+}
+
+struct JITCallGuard {
```
I feel it's probably helpful to copy the comment in your README here as well as people might copy/paste the code snippet without reading readme, e.g.:
```
To load torchscript model for mobile we need some special setup which is placed in `struct JITCallGuard` in this example. It may change in future, you can track the latest changes keeping an eye in our [pytorch android jni code]([https://github.com/pytorch/pytorch/blob/master/android/pytorch_android/src/main/cpp/pytorch_jni_jit.cpp#L28)
```

also, maybe change "need some special setup" to a bit more context, e.g. "need set these guards because mobile build doesn't support features like autograd for smaller build size"

IvanKobzarev committed:
Update on "[android] test_app example linking to pytorch_android aar content"


Example of using direct linking to pytorch_jni library from aar and updating android/README.md with the tutorial how to do it.

Adding `nativeBuild` dimension to `test_app`, using direct aar dependencies, as headers packaging is not landed yet, excluding `nativeBuild` from building by default for CI.

Additional change to `scripts/build_pytorch_android.sh`:

Skipping clean task here as android gradle plugin 3.3.2 exteralNativeBuild has problems with it when abiFilters are specified.

Will be returned back in the following diffs with upgrading of gradle and android gradle plugin versions.

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[android] test_app example linking to pytorch_android aar content"


Example of using direct linking to pytorch_jni library from aar and updating android/README.md with the tutorial how to do it.

Adding `nativeBuild` dimension to `test_app`, using direct aar dependencies, as headers packaging is not landed yet, excluding `nativeBuild` from building by default for CI.

Additional change to `scripts/build_pytorch_android.sh`:

Skipping clean task here as android gradle plugin 3.3.2 exteralNativeBuild has problems with it when abiFilters are specified.

Will be returned back in the following diffs with upgrading of gradle and android gradle plugin versions.

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[android] test_app example linking to pytorch_android aar content"


Example of using direct linking to pytorch_jni library from aar and updating android/README.md with the tutorial how to do it.

Adding `nativeBuild` dimension to `test_app`, using direct aar dependencies, as headers packaging is not landed yet, excluding `nativeBuild` from building by default for CI.

Additional change to `scripts/build_pytorch_android.sh`:

Skipping clean task here as android gradle plugin 3.3.2 exteralNativeBuild has problems with it when abiFilters are specified.

Will be returned back in the following diffs with upgrading of gradle and android gradle plugin versions.

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[android] test_app example linking to pytorch_android aar content"


Example of using direct linking to pytorch_jni library from aar and updating android/README.md with the tutorial how to do it.

Adding `nativeBuild` dimension to `test_app`, using direct aar dependencies, as headers packaging is not landed yet, excluding `nativeBuild` from building by default for CI.

Additional change to `scripts/build_pytorch_android.sh`:

Skipping clean task here as android gradle plugin 3.3.2 exteralNativeBuild has problems with it when abiFilters are specified.

Will be returned back in the following diffs with upgrading of gradle and android gradle plugin versions.

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[android] test_app example linking to pytorch_android aar content"


Example of using direct linking to pytorch_jni library from aar and updating android/README.md with the tutorial how to do it.

Adding `nativeBuild` dimension to `test_app`, using direct aar dependencies, as headers packaging is not landed yet, excluding `nativeBuild` from building by default for CI.

Additional change to `scripts/build_pytorch_android.sh`:

Skipping clean task here as android gradle plugin 3.3.2 exteralNativeBuild has problems with it when abiFilters are specified.

Will be returned back in the following diffs with upgrading of gradle and android gradle plugin versions.

Differential Revision: [D22118945](https://our.internmc.facebook.com/intern/diff/D22118945)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[android] test_app example linking to pytorch_android aar content"


Example of using direct linking to pytorch_jni library from aar and updating android/README.md with the tutorial how to do it.

Adding `nativeBuild` dimension to `test_app`, using direct aar dependencies, as headers packaging is not landed yet, excluding `nativeBuild` from building by default for CI.

Additional change to `scripts/build_pytorch_android.sh`:

Skipping clean task here as android gradle plugin 3.3.2 exteralNativeBuild has problems with it when abiFilters are specified.

Will be returned back in the following diffs with upgrading of gradle and android gradle plugin versions.

Differential Revision: [D22118945](https://our.internmc.facebook.com/intern/diff/D22118945)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[android] test_app example linking to pytorch_android aar content"


Example of using direct linking to pytorch_jni library from aar and updating android/README.md with the tutorial how to do it.

Adding `nativeBuild` dimension to `test_app`, using direct aar dependencies, as headers packaging is not landed yet, excluding `nativeBuild` from building by default for CI.

Additional change to `scripts/build_pytorch_android.sh`:

Skipping clean task here as android gradle plugin 3.3.2 exteralNativeBuild has problems with it when abiFilters are specified.

Will be returned back in the following diffs with upgrading of gradle and android gradle plugin versions.

Differential Revision: [D22118945](https://our.internmc.facebook.com/intern/diff/D22118945)

[ghstack-poisoned]
